src/Hangfire.Tags.SqlServer/SqlTagsServiceStorage.cs

```diff
@@ -198,21 +197,21 @@ select count(*)
                 jobsSql += $"  inner join [{_options.SchemaName}].[Set] s{i} on j.Id=s{i}.Value and s{i}.[Key]=@tag{i}";
             }
 
-            jobsSql += 
+            jobsSql +=
 $@"
   where (@stateName IS NULL OR LEN(@stateName) = 0 OR j.StateName=@stateName)
 )
 select j.*, s.Reason as StateReason, s.Data as StateData
 from [{_options.SchemaName}].Job j with (nolock)
-inner join cte on cte.Id = j.Id 
-left join [{_options.SchemaName}].State s with (nolock) on j.StateId = s.Id
-where cte.row_num between @start and @end 
+inner join cte on cte.Id = j.Id
+inner join [{_options.SchemaName}].State s with (nolock) on j.StateId = s.Id and j.id = s.jobId
```

Comment:
Index on State table contains both Id and JobId.
Although this looks redundant, it reduces query cost from 950 to 0.5 with millions jobs in the table